### PR TITLE
Auto scaling

### PR DESCRIPTION
**Pull request by nickumia-reisys**
Related to https://github.com/GSA/datagov-deploy/issues/3669

Reference: https://karpenter.sh/v0.8.0/getting-started/getting-started-with-terraform/

New Additions:
- Deploy Karpenter to utilize AWS Autoscaling capabilities
- Add tags to EKS and VPC to allow Karpenter to find relevant resources
- Add IAM Policies for Karpenter to create new EC2 instances
- Add Karpenter provisioner that is used to create the new EC2 instances

Updates:
- Very important: `cluster_endpoint_private_access = true` is necessary to get Karpenter to watch resources

### CODE REVIEW

File: terraform/modules/provision-k8s/k8s-autoscaler.tf [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+          cpu: 1000
+      provider:
+        launchTemplate: <launch-template-name>
+        subnetSelector:
+          karpenter.sh/discovery: <cluster-name>
+      ttlSecondsAfterEmpty: 30
+  TEMPLATE
+}
+
+resource "null_resource" "autoscaler-provisioner" {
```
Comment (mogul) [inline terraform/modules/provision-k8s/k8s-autoscaler.tf:23]: You can replace the use of the null_resource and all the monkeying about with the kubeconfig by using [the new `kubernetes_manifest` Terraform resource](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/manifest).

Essentially you can do 
```terraform
resource "kubernetes_manifest" "karpenter-provisioner" {
  manifest = yamldecode(local.karpenter_provisioner)
}
```
That would clean this up a lot!

Comment (mogul) [reply terraform/modules/provision-k8s/k8s-autoscaler.tf:23]: ...and remove the need for more mapping outputs to inputs between the modules...

**Comment by nickumia-reisys** [comment]
There's a problem with the kubernetes manfest, it keeps giving this error and I can't figure out the cause.  The provider version seems compatible (v2.9) and the manifest is able to apply with `kubectl` so I'm not sure what the issue is.

image

**Comment by nickumia-reisys** [comment]
As to last [CI run](https://github.com/GSA/datagov-brokerpak-eks/runs/5817715417?check_suite_focus=true#step:6:34), it seems like [this is awfully suspicious](https://stackoverflow.com/a/67540365).

**Review by mogul — APPROVED** [review]

**Comment by mogul** [comment]
Ready to remove draft status?

**Comment by mogul** [comment]
I think the answer will ultimately be to [put things that create CRDs in a separate module from things that use those CRDs](https://github.com/hashicorp/terraform-provider-kubernetes/issues/1380#issuecomment-933520161), which will be possible once [this CSB change allows us to use directory paths naturally in the brokerpak HCL](https://github.com/cloudfoundry/cloud-service-broker/issues/114#issuecomment-1086251171).

Until then, I think we should try [this other workaround which wraps the CRD usage in a generic `helm_release`](https://github.com/hashicorp/terraform-provider-kubernetes/issues/1380#issuecomment-962058148). (That way we don't have to revert back to passing values between our modules.)

**Review by mogul — APPROVED** [review]

**Pull request merged**